File: 131-palindrome-partitioning/131-palindrome-partitioning.cpp
class Solution {
public:
    
    void solve(vector<vector<string>> &ans,vector<string> &path,string s,int index)
    {
        if(index==s.size())
        {
            ans.push_back(path);
            return;
        }
        for(int i=index;i<s.size();i++)
        {
            if(isPalindrome(s,index,i))
            {
                path.push_back(s.substr(index,i-index+1));
                solve(ans,path,s,i+1);
                path.pop_back();
            }
        }
    }
    
    bool isPalindrome(string s,int start,int end)
    {
        if(start<=end)
        {
            if(s[end]!=s[start]) return false;
            return isPalindrome(s,++start,--end);
        }
        else return true;
    }
    
    vector<vector<string>> partition(string s) {
        vector<vector<string> > ans;
        vector<string> path;
        solve(ans,path,s,0);
        return ans;
    }
};

File: 131-palindrome-partitioning/131-palindrome-partitioning.java
class Solution {
    
    
    void solve(List<List<String> > ans,List<String> path,String s,int index)
    {
        if(index==s.length())
        {
            ans.add(new ArrayList(path));
            return;
        }
        for(int i=index;i<s.length();i++)
        {
            if(isPalindrome(s,index,i))
            {
                path.add(s.substring(index,i+1));
                solve(ans,path,s,i+1);
                path.remove(path.size()-1);
            }
        }
    }
    
    boolean isPalindrome(String s,int start,int end)
    {
        if(start<=end)
        {
            if(s.charAt(start)!=s.charAt(end)) return false;
            return isPalindrome(s,++start,--end);
        }
        return true;
    }
    
    
    public List<List<String>> partition(String s) {
        List<List<String>> ans=new ArrayList<List<String> >();
        List<String> path=new ArrayList<String>();
        solve(ans,path,s,0);
        return ans;
    }
}